File: src/challenges/python/problems/game_of_life_test.py
# https://leetcode.com/problems/game-of-life

from core.problem_base import *

class GameOfLife(ProblemBase):
    def Solution(self, board: list[list[int]]) -> list[list[int]]:
        """
        Do not return anything, modify board in-place instead.
        """
        directions = [
            [ -1, -1 ], [ -1, 0 ], [ -1, 1 ],
            [  0, -1 ],            [  0, 1 ],
            [  1, -1 ], [  1, 0 ], [  1, 1 ],
        ]

        count = len(directions)
        rows = len(board)
        cols = len(board[0])

        for row in range(rows):
            for col in range(cols):
                neighbors = 0

                for i in range(count):
                    y = row + directions[i][0]
                    x = col + directions[i][1]

                    if y >= 0 and y < rows \
                        and x >= 0 and x < cols \
                        and (board[y][x] == 1 or board[y][x] >= 20):
                        neighbors += 1

                board[row][col] = neighbors + (10 if board[row][col] == 0 else 20)

        for row in range(rows):
            for col in range(cols):
                neighbors = board[row][col] - (20 if board[row][col] >= 20 else 10)

                if neighbors < 2 or neighbors > 3:
                    board[row][col] = 0
                elif neighbors == 3 and board[row][col] < 20:
                    board[row][col] = 1
                elif board[row][col] >= 20:
                    board[row][col] = 1
                else:
                    board[row][col] = 0
        
        return board

if __name__ == '__main__':
    TestGen(GameOfLife) \
        .Add(lambda tc: tc.Param([[0,1,0],[0,0,1],[1,1,1],[0,0,0]]).Result([[0,0,0],[1,0,1],[0,1,1],[0,1,0]])) \
        .Add(lambda tc: tc.Param([[1,1],[1,0]]).Result([[1,1],[1,1]])) \
        .Run()

File: src/rust/challenges/src/problems/game_of_life_test.rs
// https://leetcode.com/problems/game-of-life

pub fn solution1(board: &mut Vec<Vec<i32>>) -> Vec<Vec<i32>> {
    let directions = vec![
        vec![-1, -1], vec![-1, 0], vec![-1, 1],
        vec![0, -1], vec![0, 1],
        vec![1, -1], vec![1, 0], vec![1, 1],
    ];

    let count = directions.len();
    let rows = board.len();
    let cols = board[0].len();

    for row in 0..rows {
        for col in 0..cols {
            let mut neighbors = 0;

            for i in 0..count {
                let y = row as i32 + directions[i][0];
                let x = col as i32 + directions[i][1];

                if y >= 0 && y < rows as i32 && x >= 0 && x < cols as i32 &&
                    (board[y as usize][x as usize] == 1 || board[y as usize][x as usize] >= 20) {
                    neighbors += 1;
                }
            }

            if board[row][col] == 0 {
                board[row][col] = neighbors + 10;
            } else {
                board[row][col] = neighbors + 20;
            }
        }
    }

    for row in 0..rows {
        for col in 0..cols {
            let mut neighbors = 0;
            if board[row][col] >= 20 {
                neighbors = board[row][col] - 20;
            } else {
                neighbors = board[row][col] - 10;
            }

            if neighbors < 2 || neighbors > 3 {
                board[row][col] = 0;
            } else if neighbors == 3 && board[row][col] < 20 {
                board[row][col] = 1;
            } else if board[row][col] >= 20 {
                board[row][col] = 1;
            } else {
                board[row][col] = 0;
            }
        }
    }

    board
}